aQute.libg/test/aQute/libg/uri/URIUtilsTest.java

```diff
@@ -1,56 +1,125 @@
-		assertEquals("http://example.com/bar.xml", result.toString());
+		assertThat(result).hasToString("http://example.com/bar.xml");
 	}
 
+	@Test
 	public void testResolveBlank() throws Exception {
 		URI result = URIUtil.resolve(URI.create("http://example.com/foo.xml"), "");
-		assertEquals("http://example.com/foo.xml", result.toString());
+		assertThat(result).hasToString("http://example.com/foo.xml");
 	}
 
+	@Test
 	public void testResolveFragmentBlank() throws Exception {
 		URI result = URIUtil.resolve(URI.create("http://example.com/foo.xml#bar"), "");
-		assertEquals("http://example.com/foo.xml", result.toString());
+		assertThat(result).hasToString("http://example.com/foo.xml");
 	}
 
+	@Test
+	@EnabledOnOs(WINDOWS)
 	public void testResolveAbsoluteWindowsPath() throws Exception {
-		if (IO.isWindows()) {
-			URI result = URIUtil.resolve(URI.create("file:/C:/Users/jimbob/base.txt"), "C:\\Users\\sub dir\\foo.txt");
-			assertEquals("file:/C:/Users/sub%20dir/foo.txt", result.toString());
-			result = URIUtil.resolve(URI.create("file:/C:/Users/jimbob/base.txt"), "C:/Users/sub dir/bar.txt");
-			assertEquals("file:/C:/Users/sub%20dir/bar.txt", result.toString());
-		}
+		URI result = URIUtil.resolve(URI.create("file:/C:/Users/jimbob/base.txt"), "C:\\Users\\sub dir\\foo.txt");
+		assertThat(result).hasToString("file:/C:/Users/sub%20dir/foo.txt");
+		result = URIUtil.resolve(URI.create("file:/C:/Users/jimbob/base.txt"), "C:/Users/sub dir/bar.txt");
+		assertThat(result).hasToString("file:/C:/Users/sub%20dir/bar.txt");
 	}
 
+	@Test
+	@EnabledOnOs(WINDOWS)
 	public void testResolveRelativeWindowsPath() throws Exception {
-		if (IO.isWindows()) {
-			URI result = URIUtil.resolve(URI.create("file:/C:/Users/jim/base.txt"), "sub dir\\foo.txt");
-			assertEquals("file:/C:/Users/jim/sub%20dir/foo.txt", result.toString());
-		}
+		URI result = URIUtil.resolve(URI.create("file:/C:/Users/jim/base.txt"), "sub dir\\foo.txt");
+		assertThat(result).hasToString("file:/C:/Users/jim/sub%20dir/foo.txt");
 	}
 
+	@Test
+	@EnabledOnOs(WINDOWS)
 	public void testResolveUNCWindowsPath() throws Exception {
-		if (IO.isWindows()) {
-			URI result = URIUtil.resolve(URI.create("file:/C:/Users/jim/base.txt"), "\\\\server\\share\\foo.txt");
-			assertEquals("file:////server/share/foo.txt", result.toString());
-			result = URIUtil.resolve(URI.create("file:/C:/Users/jim/base.txt"), "//server/share/foo.txt");
-			assertEquals("file:////server/share/foo.txt", result.toString());
-		}
+		URI result = URIUtil.resolve(URI.create("file:/C:/Users/jim/base.txt"), "\\\\server\\share\\foo.txt");
+		assertThat(result).hasToString("file:////server/share/foo.txt");
+		result = URIUtil.resolve(URI.create("file:/C:/Users/jim/base.txt"), "//server/share/foo.txt");
+		assertThat(result).hasToString("file:////server/share/foo.txt");
+	}
+
+	private static Stream<Arguments> fromURI_returnsPath_platformIndependent() {
```

Comment:
I must say this is some seriously unreadable stuff. I can see that it is powerful but it makes it impossible to set a break point on the one assert data that you want to debug. So I would rather not do this and have the test method just do `assertThat(URIUtil.fromURI("uri")).isEqualTo("expected");` repeatedly. Then I can put a break point on whatever line I want to debug that test data.

And then we don't need Junit 5 for this PR.

Reply:
This was the primary reason for me upgrading to JUnit 5, and now you want me to abandon it! 😆

> I must say this is some seriously unreadable stuff.

I'd like a bit more clarity on what you mean by "unreadable". Originally, I had the Arguments stream laid out quite a bit neater, in a more readable, tabular way, but the Eclipse formatting rules turned it into the mess that you currently see. Would fixing the formatting fix this objection? Or is there more to it?

> I can see that it is powerful but it makes it impossible to set a break point on the one assert data that you want to debug. 

I think this is worth a bit of a discussion, because as it currently stands a practical consequence of this objection is a blanket ban on the use of parameterized tests. But I get what you are saying about debugging a single result. Does the following alternative workflow work for you:

1. Set your breakpoint on the parameterized test method.
2. Run the test in the IDE.
3. Look at the results in the JUnit tab, find the parameter combination that you'd like to debug.
4. Right-click on the result for that parameter combination, and click "Debug".

The Eclipse JUnit runner will now re-run the test for that specific parameter combination, and it will stop at your breakpoint.

I've gone ahead and made the change you requested anyway, but still interested in your thoughts on the above for future reference.

Reply:
In general, I like test case methods to be simple, A then B then C. Easy to read, easy to debug, easy to change. It may mean some more code repetition but it is a test case not production code. Sometime extracting common code to a function does make sense to be shared among test methods. But in this case the common code was a single assert. So there it little saved by sharing.

I have seen test cases which are so complicated that no one left on the project understands them or how to modify, extends or fix them. So I am very much of the opinion that test cases should be plain dead simple.

This example took advantage of a JUnit 5 feature to derive multiple test methods from a single test method and a stream of data input. For this case, it seems unnecessarily complicated for what we need to do. You could also have softly asserted all the asserts to get the maximum results if you wanted to avoid an early failure stopping the remaining asserts.

Reply:
I like test methods to be as simple as possible too. But was the parameterized test really that complicated? I'd agree that it was a little unreadable (primarily due to the bad formatting), but conceptually it's quite simple - a single test method (with a single assert), and a parameter list. However, I agree that in this case with a single assert there wasn't a lot of boilerplate saved - in fact, probably when you look at the clutter introduced by `Arguments.of` it was even a net loss! I could have used more compact ways of doing the parameter list, but I've found them to be less type safe (I'm still learning here myself).

You're right that soft assertions can be used to prevent crashing out the test on the first assertion (this is what the latest push looks like). However, I think there are still some reasons to prefer parameterized approach:
1. The output in Eclipse is IMHO more easily navigable. The results are hierarchically grouped, one line per test. You can see which one(s) failed and click on it/them for the stack trace. The output of multiple soft assertion failures can be a bit difficult to follow if there are many failures.
2. Soft assertions prevent an assertion short-circuiting the test, but there's nothing to stop an uncaught exception from short-circuiting the test. Each run of a parameterized test is completely isolated from the others.